src/org/oidc/service/AbstractService.java

```diff
@@ -151,11 +178,64 @@ public abstract void updateServiceContext(
      * @param stateKey the key that identifies the State object
      * @return the parsed and to some extent verified response
      **/
-    public Message parseResponse(
-            String responseBody,
-            SerializationType serializationType,
-            String stateKey) {
-        return null;
+    public Message parseResponse(String responseBody, SerializationType serializationType, String stateKey) throws Exception {
+        if(serializationType == null) {
+            serializationType = this.serializationType;
+        }
+
+        if(SerializationType.URL_ENCODED.equals(serializationType)) {
+            responseBody = ServiceUtil.getUrlQueryReference(responseBody);
```

Comment:
You are retrieving query or reference from responseBody. Why are you assigning it back to responseBody?

Reply:
created a new var called urlInfo